src/dune_pkg/archive_driver.ml

```diff
@@ -0,0 +1,106 @@
+open Stdune
+module Process = Dune_engine.Process
+open Fiber.O
+
+module Command = struct
+  type t =
+    { bin : Path.t
+    ; args : archive:Path.t -> target_in_temp:Path.t -> string list
+    }
+end
+
+type t =
+  { command : Command.t Lazy.t
+  ; suffixes : string list
+  }
+
+let which bin_name = Bin.which ~path:(Env_path.path Env.initial) bin_name
+
+let tar_args ~archive ~target_in_temp =
+  [ "xf"; Path.to_string archive; "-C"; Path.to_string target_in_temp ]
+;;
+
+let zip_args ~archive ~target_in_temp =
+  [ Path.to_string archive; "-d"; Path.to_string target_in_temp ]
+;;
+
+let tar =
+  let command =
+    lazy
+      (match
+         (* Test for tar before bsdtar as tar is more likely to be installed
+            and both work equally well for tarballs. *)
+         List.find_map [ "tar"; "bsdtar" ] ~f:which
+       with
+       | Some bin -> { Command.bin; args = tar_args }
+       | None -> Dune_engine.Utils.program_not_found "tar" ~loc:None)
+  in
+  { command; suffixes = [ ".tar"; ".tar.gz"; ".tgz"; ".tar.bz2"; ".tbz" ] }
+;;
+
+let zip =
+  let command =
+    lazy
+      (match which "unzip" with
+       | Some bin -> { Command.bin; args = zip_args }
+       | None ->
+         (* Fall back to using tar to extract zip archives, which is possible in some cases. *)
+         (match
+            (* Test for bsdtar before tar, as if bsdtar is installed then it's
+               likely that the tar binary is GNU tar which can't extract zip
+               archives, whereas bsdtar can. If bsdtar is absent, try using the
+               tar command anyway, as on BSD, MacOS, and Windows systems, the tar
+               command can extract zip archives. *)
+            List.find_map [ "bsdtar"; "tar" ] ~f:which
+          with
+          | Some bin -> { Command.bin; args = tar_args }
```

Comment:
Do you know what would be the error if you try to use `GNUtar` to unzip, and it fails because you are on Linux? I'm asking because it might sound weird to users to get a `tar` error when `unzipping` :confused:

Reply:
I did an experiment to test this and the error was:
```
File "dune.lock/tsort.pkg", line 10, characters 7-65:
10 |   (url https://github.com/dmbaturin/ocaml-tsort/archive/2.0.0.zip)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error:
failed to unpack archive downloaded from
https://github.com/dmbaturin/ocaml-tsort/archive/2.0.0.zip
reason:
unable to extract "_build/.temp/dune_d7c5e3_2.0.0.zip/download"
```

It's not great in that it doesn't help the user figure out how to fix the problem (ie. install `unzip`) but it's no worse than the current error message (it is just the current error message). You'd get the same error if `tar` wasn't installed. We should improve this but not in this PR.

Reply:
I've created https://github.com/ocaml/dune/issues/11544 for improving error messages. Might be a good first issue for an outreachy intern to tackle.

Reply:
Thanks for opening the issue.  My main concern was that the message was stating that `tar` is the problem with a zip file. As the error message is generic, that's fine :ok_hand: